pandas/core/frame.py

```diff
@@ -6892,56 +6892,93 @@ def cov(self, min_periods=None):
         """
-        Compute pairwise correlation between rows or columns of two DataFrame
-        objects.
+        Compute correlation between rows or columns of two DataFrame objects
+        or DataFrame and Series objects.
 
         Parameters
         ----------
         other : DataFrame, Series
         axis : {0 or 'index', 1 or 'columns'}, default 0
             0 or 'index' to compute column-wise, 1 or 'columns' for row-wise
+        how : {'pairwise', 'all'}, default 'pairwise' if other is a DataFrame
+            and default 'all' if other is a Series object
+            * 'pairwise' to compute correlation between rows or columns with
+              the same name/index,
+            * 'all' to compute correlation between all possible pairs of rows
+              or columns
         drop : boolean, default False
             Drop missing indices from result, default returns union of all
+        method : {'pearson', 'kendall', 'spearman'}
+            * pearson : standard correlation coefficient
+            * kendall : Kendall Tau correlation coefficient
+            * spearman : Spearman rank correlation
+        min_periods : int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result. Currently only available for pearson
+            and spearman correlation
+
+        Returns
+        -------
+        y : DataFrame
 
         Returns
         -------
-        correls : Series
+        correl : Series (how='pairwise') or DataFrame (how='all')
         """
+
         axis = self._get_axis_number(axis)
         this = self._get_numeric_data()
 
         if isinstance(other, Series):
-            return this.apply(other.corr, axis=axis)
+            return this.apply(other.corr, axis=axis,
+                              method=method, min_periods=min_periods)
 
-        other = other._get_numeric_data()
+        if not isinstance(other, DataFrame):
+            raise TypeError("'other' parameter should be a DataFrame "
+                            "or a Series object")
 
-        left, right = this.align(other, join='inner', copy=False)
+        other = other._get_numeric_data()
 
-        # mask missing values
-        left = left + right * 0
-        right = right + left * 0
+        if isinstance(this.columns, MultiIndex) \
+                or isinstance(other.columns, MultiIndex):
+            raise TypeError("MultiIndex is not supported")
 
         if axis == 1:
-            left = left.T
-            right = right.T
-
-        # demeaned data
-        ldem = left - left.mean()
-        rdem = right - right.mean()
+            this = this.transpose()
+            other = other.transpose()
+
+        if how == 'all':
+            corr = np.zeros((this.shape[1], other.shape[1]))
+
+            for i, col1 in enumerate(this.columns):
+                for j, col2 in enumerate(other.columns):
+                    corr[i, j] = this.loc[:, col1].corr(
```

Comment:
dont you just to do a single triangle? (and copy to the other)?

use ``this.iloc[:, j]`` and so on. test this with non-uniques columns (you could raise on non-uniques alternatively)

Reply:
but still use the iloc pattern here

Reply:
pls fix this

Reply:
There is no triangle since columns and rows have different indexes (the case for `how='all'`)
Changed `.loc()` to `.iloc()`